include/envoy/matcher/matcher.h

```diff
@@ -0,0 +1,214 @@
+
+  // This encodes three states:
+  // Not enough data to complete the match: {false, {}}
+  // Completed the match, no match: {true, {}}
+  // Completed the match, match: {true, on_match}
+  struct MatchResult {
+    const MatchState match_state_;
+    const absl::optional<OnMatch<DataType>> on_match_;
+  };
+
+  // Attempts to match against the matching data (which should contain all the data requested via
+  // matching requirements). If the match couldn't be completed, {false, {}} will be returned.
+  // If a match result was determined, {true, action} will be returned. If a match result was
+  // determined to be no match, {true, {}} will be returned.
+  virtual MatchResult match(const DataType& matching_data) PURE;
+};
+
+// InputMatcher provides the interface for determining whether an input value matches.
+class InputMatcher {
+public:
+  virtual ~InputMatcher() = default;
+
+  /**
+   * Whether the provided input is a match.
+   * @param absl::optional<absl::string_view> the value to match on. Will be absl::nullopt if the
+   * lookup failed.
+   */
+  virtual bool match(absl::optional<absl::string_view> input) PURE;
+};
+
+using InputMatcherPtr = std::unique_ptr<InputMatcher>;
+
+/**
+ * Factory for registering custom input matchers.
+ */
+class InputMatcherFactory : public Config::TypedFactory {
+public:
+  virtual InputMatcherPtr createInputMatcher(const Protobuf::Message& config) PURE;
+
+  std::string category() const override { return "envoy.matching.input_matcher"; }
+};
+
+struct DataInputGetResult {
+  enum class DataAvailability {
+    // The data is not yet available.
+    NotAvailable,
+    // Some data is available, but more might arrive.
+    MoreDataMightBeAvailable,
+    // All the data is available.
+    AllDataAvailable
+  };
+
+  DataAvailability data_availability_;
+  // The resulting data. This will be absl::nullopt if we don't have sufficient data available (as
+  // per data_availability_) or because no value was extracted. For example, consider a DataInput
+  // which attempts to look a key up in the map: if we don't have access to the map yet, we return
+  // absl::nullopt with NotAvailable. If we have the entire map, but the key doesn't exist in the
+  // map, we return absl::nullopt with AllDataAvailable.
+  absl::optional<absl::string_view> data_;
+
+  friend std::ostream& operator<<(std::ostream& out, const DataInputGetResult& result) {
+    out << "data input: " << (result.data_ ? result.data_.value() : "n/a");
+    switch (result.data_availability_) {
+    case DataInputGetResult::DataAvailability::NotAvailable:
+      out << " (not available)";
+      break;
+    case DataInputGetResult::DataAvailability::MoreDataMightBeAvailable:
+      out << " (more data available)";
+      break;
+    case DataInputGetResult::DataAvailability::AllDataAvailable:;
+    }
+    return out;
+  }
+};
+
+/**
+ * Interface for types providing a way to extract a string from the DataType to perform matching on.
```

Comment:
I thought we supported matchers that were not always operating on strings as an extension with https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/common/matcher/v3/matcher.proto#L65. Are we saying even these reduce down to strings?

Reply:
Yeah my understanding from the last conversations with @markdroth was that the inputs would always be strings and the matchers would be responsible for converting that into other formats: e.g. an any matcher could split on a character and have a delegate matcher it uses to match the individual pieces against. 

I think it would be neat to have typed inputs but you end up with a fair bit of complexity around making sure that the input and matcher is compatible, so I'm not sure if its worth the complexity?

Reply:
It is worth noting that the question of whether the matching infrastructure can handle types other than string doesn't actually affect the API.  I think it's probably fine to start with an implementation that currently handles only strings, and if we later need to change it such that the input plugins and matcher plugins are type-specific (and must both use the same type for a given match), we can do that as an implementation change without affecting the API.